pkgs/development/lisp-modules/default.nix

```diff
@@ -0,0 +1,299 @@
+      # systems on run time.
+      #
+      # Also useful when the pname is differrent than the system name,
+      # such as when using reverse domain naming.
+      systems ? [ pname ],
+
+      # The .asd files that this package provides
+      # TODO(kasper): remove
+      asds ? systems,
+
+      # Other args to mkDerivation
+      ...
+    } @ args:
+
+    stdenv.mkDerivation (rec {
+      inherit
+        pname version nativeLibs javaLibs lispLibs systems asds
+        pkg program flags faslExt
+      ;
+
+      # When src is null, we are building a lispWithPackages and only
+      # want to make use of the dependency environment variables
+      # generated by build-asdf-system
+      dontUnpack = src == null;
+
+      # Portable script to build the systems.
+      #
+      # `lisp` must evaluate this file then exit immediately. For
+      # example, SBCL's --script flag does just that.
+      #
+      # NOTE:
+      # Every other library worked fine with asdf:compile-system in
+      # buildScript.
+      #
+      # cl-syslog, for some reason, signals that CL-SYSLOG::VALID-SD-ID-P
+      # is undefined with compile-system, but works perfectly with
+      # load-system. Strange.
+
+      # TODO(kasper) portable quit
+      asdfFasl = buildAsdf { inherit asdf pkg program flags faslExt; };
+
+      buildScript = substituteAll {
+        src = ./builder.lisp;
+        asdf = "${asdfFasl}/asdf.${faslExt}";
+      };
+
+      preConfigure = ''
+        source ${./setup-hook.sh}
+        buildAsdfPath
+      '';
+
+      buildPhase = optionalString (src != null) ''
+        export CL_SOURCE_REGISTRY=$CL_SOURCE_REGISTRY:$src//
+        export ASDF_OUTPUT_TRANSLATIONS="$src:$(pwd):${storeDir}:${storeDir}"
+        ${pkg}/bin/${program} ${flags} < $buildScript
+      '';
+
+      # Copy compiled files to store
+      #
+      # Make sure to include '$' in regex to prevent skipping
+      # stuff like 'iolib.asdf.asd' for system 'iolib.asd'
+      #
+      # Same with '/': `local-time.asd` for system `cl-postgres+local-time.asd`
+      installPhase =
+        let
+          mkSystemsRegex = systems:
+            concatMapStringsSep "\\|" (replaceStrings ["." "+"] ["[.]" "[+]"]) systems;
+        in
+      ''
+        mkdir -pv $out
+        cp -r * $out
+
+        # Remove all .asd files except for those in `systems`.
+        find $out -name "*.asd" \
+        | grep -v "/\(${mkSystemsRegex systems}\)\.asd$" \
+        | xargs rm -fv || true
+      '';
+
+      dontPatchShebangs = true;
+
+      # Not sure if it's needed, but caused problems with SBCL
+      # save-lisp-and-die binaries in the past
+      dontStrip = true;
+
+    } // (args // {
+      src = if builtins.length (args.patches or []) > 0
+            then pkgs.applyPatches { inherit (args) src patches; }
+            else args.src;
+      patches = [];
+      propagatedBuildInputs = args.propagatedBuildInputs or []
```

Comment:
Don't they default to empty lists?

Reply:
I should check if it will work with the line rem oved

Reply:
Yes, patches are empty by default, but it also appears in args which is merged into the final attrset. We want to applyPatches manually and set it to an empty list afterwards

for `propagatedBuildInputs`, It was part of this: https://github.com/NixOS/nixpkgs/pull/202726 @mmarx do you remember why it was needed?

Reply:
If I recall correctly, there were some systems that need runtime dependencies present at load time (they call `trivial-package-manager`'s `ensure-program` as part of the `load-op`), and propagated build inputs from such packages were not propagated to a `lispWithPackages`.